[Clang][AArch64]Add FP8 ACLE macros implementation

CarolineConcatto opened the pull request:
This patch implements the macros described in the ACLE[1]

[1] https://github.com/ARM-software/acle/blob/main/main/acle.md#modal-8-bit-floating-point-extensions

CarolineConcatto committed:
[Clang][AArch64]Add FP8 ACLE macros implementation

This patch implements the macros described in the ACLE[1]

[1] https://github.com/ARM-software/acle/blob/main/main/acle.md#modal-8-bit-floating-point-extensions

paulwalker-arm commented on clang/lib/Basic/Targets/AArch64.cpp:
```diff
@@ -1046,6 +1082,29 @@ bool AArch64TargetInfo::handleTargetFeatures(std::vector<std::string> &Features,
+    if (Feature == "+fp8dot4")
+      HasFP8DOT4 = true;
+    if (Feature == "+ssve-fp8dot2")
+      HasSSVE_FP8DOT2 = true;
```
This and the other `SSVE` features should also set `HasSME2`?

paulwalker-arm commented on clang/lib/Basic/Targets/AArch64.cpp:
```diff
@@ -1046,6 +1082,29 @@ bool AArch64TargetInfo::handleTargetFeatures(std::vector<std::string> &Features,
+
+    if (Feature == "+fp8")
+      HasFP8 = true;
+    if (Feature == "+fp8fma")
```
`fp8...` feature extensions should also set `HasFP8`?

CarolineConcatto replied:
I could. 
I am not sure I should reproduce the same we have in  lib/Target/AArch64/AArch64Features.td also here. 
Is that what you  are suggestion here?
At first glance it does not look like this is a representation of what we have in 
 lib/Target/AArch64/AArch64Features.td

paulwalker-arm replied:
I don't fully understand the effect of setting these bools and how the dependencies in `AArch64Features.td` correspond to setting the feature macros.  What I'm mainly after is consistency and so for the SME2 extensions you're setting `HasSME2` as well.  So I guess the question is why that is necessary?  If it's not, then perhaps the way to go is to remove those rather than adding the ones I suggest.

The most important part to verify is that all the feature macros a user would expect to be defined by a specific +feat are in fact defined.  If that is managed by `AArch64Features.td` then great, if not then that would explain why the key feature dependencies are duplicated here.

CarolineConcatto replied:
If I run -E I can see all the macros dependency defined without adding them here. So I think it is fine to only set the new feature.
Maybe we need to look to the other definitions and check if we need to clean them up too.

paulwalker-arm commented on clang/test/Preprocessor/aarch64-target-features.c:
```diff
@@ -744,3 +744,34 @@
+// CHECK-SSVE-FP8FMA: __ARM_FEATURE_SSVE_FP8FMA 1
+
+// RUN: %clang --target=aarch64 -march=armv9-a+sme-f8f32 -x c -E -dM %s -o - | FileCheck --check-prefix=CHECK-SMEF8F32 %s
+// CHECK-SMEF8F32: __ARM_FEATURE_LOCALLY_STREAMING 1
```
Are the two `__ARM_FEATURE_LOCALLY_STREAMING` checks relevant for the tests?

CarolineConcatto replied:
Do you mean for the sme-f8f32 and sme-f8f16? If so I was just reproducing what was done for sme-b16b16 and sme-f16f16.

paulwalker-arm replied:
Ok, in which case this is likely an extension of the previous point.  If the feature relationships have to be duplicated then it makes sense we verify `__ARM_FEATURE_LOCALLY_STREAMING` here.  If they don't and `AArch64Features.td` ensures the expected SME feature flags get defined, then we don't need to reverify them here.

CarolineConcatto committed:
Remove redundancy set in handleTargetFeatures

It looks like other features are set by default, following the
dependendy in Features. We only need to set the new ones.

paulwalker-arm commented on clang/test/Preprocessor/aarch64-target-features.c:
```diff
@@ -744,3 +744,38 @@
+// CHECK-FP8DOT4: __ARM_FEATURE_FP8 1
+// CHECK-FP8DOT4: __ARM_FEATURE_FP8DOT4 1
+
+// RUN: %clang --target=aarch64 -march=armv9-a+ssve-fp8dot2 -x c -E -dM %s -o - | FileCheck --check-prefix=CHECK-SSVE-FP8DOT2 %s
+// CHECK-SSVE-FP8DOT2: __ARM_FEATURE_SME2 1
+// CHECK-SSVE-FP8DOT2: __ARM_FEATURE_SSVE_FP8DOT2 1
+
+// RUN: %clang --target=aarch64 -march=armv9-a+ssve-fp8dot4 -x c -E -dM %s -o - | FileCheck --check-prefix=CHECK-SSVE-FP8DOT4 %s
+// CHECK-SSVE-FP8DOT4: __ARM_FEATURE_SME2 1
+// CHECK-SSVE-FP8DOT4: __ARM_FEATURE_SSVE_FP8DOT4 1
+
+// RUN: %clang --target=aarch64 -march=armv9-a+ssve-fp8fma -x c -E -dM %s -o - | FileCheck --check-prefix=CHECK-SSVE-FP8FMA %s
+// CHECK-SSVE-FP8FMA: __ARM_FEATURE_SME2 1
+// CHECK-SSVE-FP8FMA: __ARM_FEATURE_SSVE_FP8FMA 1
```
Can the three `_ssve-fp8...` options also check for `__ARM_FEATURE_FP8`?

paulwalker-arm reviewed (commented):
Last question but otherwise this looks good to me.

CarolineConcatto committed:
Improve test output to also show FP8 macro is also set
